### ISSUE TITLE
Potential Data Leakage 

### ISSUE THREAD

**Issue by smangold459** [learn-co-curriculum/dsc-knn-with-sklearn-lab#1]
Hello,

I wanted to raise an issue to note on a potential error in the order of operations in this notebook's pre-processing section. The concern is that the notebook instantiates a OneHotEncoder and fit/transforms the entire dataset prior to performing a train_test_split. Conceptually our instructor has advised us this is a potential cause of data leakage that OHE is programmed to solve. 

If the entire dataset is fitted to the OHE there is a chance that when the train test split is performed afterwards there will potentially be entire columns in the training data that do not contain values, but they we're dummied out when the set was whole. That is, if a train test split is performed and your training data is not completely ignorant of one column that was potentially split out of the data when the test set was created, should it not perform analysis on just the columns contained in the training set?

The way we're being instructed to perform the encoding is as follows:
```
ohe = OneHotEncoder()
ohe.fit_transform(X_train)
ohe.transform(X_test)
```
Your feedback on the why is does or does not matter with respect to data leakage would be greatly appreciated :)

**Comment by smangold459** [comment]
Or perhaps does it not matter? Since if the situation I described above did occur the model would in that instance be training on a column full of zeroes and lead to a coefficient of 0 which would then nullify and test values that came through in the analysis?

**Comment by bpurdy-ds** [comment]
@smangold459 Unhelpfully, perhaps, both can be done with justification. You are correct that we are generally taught to split first to avoid data leakage. If we preprocess first, this is certainly easier since there is less code since you don't have to do everything twice. The "danger" of splitting first is that the test data by sheer bad luck of the split is missing a category type that is in the test set. Of course, since the data is supposed to be representative this case should not happen.

All that being said, we'll keep the order as it is, but I'll add a note in the lab that says why we're doing the encoding when we are.

**Comment by bpurdy-ds** [comment]
Closed the commit as described above: https://github.com/learn-co-curriculum/dsc-knn-with-scikit-learn-lab/issues/1#issue-951682636
